internal/dcache/file_manager/file_manager.go

```diff
@@ -0,0 +1,361 @@
+				chunk.Len, getChunkStartOffsetFromFileOffset(offset, &file.FileMetadata.FileLayout)))
+
+		if chunk.Len == file.FileMetadata.FileLayout.ChunkSize {
+			// Schedule the upload
+			// todo: This not always true. if some writes to this chunk were skipped then there should be a
+			// way to stage this block.
+			scheduleUpload(chunk, file)
+		}
+	}
+	common.Assert(offset == endOffset, fmt.Sprintf("Write is not successful, expected Endoffset : %d, actual EndOffset : %d", endOffset, offset))
+	common.Assert(bufOffset == len(buf), fmt.Sprintf("Amount of Bytes copied : %d is not equal to buf len %d", bufOffset, len(buf)))
+	file.lastWriteOffset = offset
+	return nil
+}
+
+// Sync Buffers for the file with dcache/azure
+// This call can come when user application calls fsync()/close()
+func (file *DcacheFile) SyncFile() error {
+	log.Debug("DistributedCache[FM]::SyncFile : Sync File for %s", file.FileMetadata.Filename)
+	var err error
+	file.StagedChunks.Range(func(chunkIdx any, Ichunk any) bool {
+		chunk := Ichunk.(*StagedChunk)
+		scheduleUpload(chunk, file)
+		err = <-chunk.Err
+		if err != nil {
+			return false
+		}
+		return true
+	})
+	return err
+}
+
+// Release all allocated buffers for the file
+func (file *DcacheFile) ReleaseFile() error {
+	log.Debug("DistributedCache[FM]::ReleaseFile :Releasing buffers for File for %s", file.FileMetadata.Filename)
+	file.StagedChunks.Range(func(chunkIdx any, Ichunk any) bool {
+		chunk := Ichunk.(*StagedChunk)
+		file.releaseChunk(chunk)
+		return true
+	})
+	return nil
+}
+
+// Get's the existing chunk from the chunks
+// or Create a new one and add it to the the chunks
+func (file *DcacheFile) getChunk(chunkIdx int64) (*StagedChunk, bool, error) {
+	log.Debug("DistributedCache::getChunk : getChunk for chunkIdx : %d, file : %s", chunkIdx, file.FileMetadata.Filename)
+	if chunkIdx < 0 {
+		return nil, false, errors.New("ChunkIdx is less than 0")
+	}
+	if chunk, ok := file.StagedChunks.Load(chunkIdx); ok {
+		return chunk.(*StagedChunk), true, nil
+	}
+	chunk, err := NewChunk(chunkIdx, file)
+	if err != nil {
+		return nil, false, err
+	}
+	file.StagedChunks.Store(chunkIdx, chunk)
+	return chunk, false, nil
+}
+
+func (file *DcacheFile) getChunkForRead(chunkIdx int64) (*StagedChunk, error) {
+	log.Debug("DistributedCache::getChunkForRead : getChunk for Read chunkIdx : %d, file : %s", chunkIdx, file.FileMetadata.Filename)
+	chunk, loaded, err := file.getChunk(chunkIdx)
+	if err == nil && !loaded {
+		close(chunk.ScheduleUpload)
+		chunk.Len = getChunkSize(chunkIdx*file.FileMetadata.FileLayout.ChunkSize, file)
+	}
+	return chunk, err
+}
+
+func (file *DcacheFile) getChunkForWrite(chunkIdx int64) (*StagedChunk, error) {
+	log.Debug("DistributedCache::getChunkForWrite : getChunk for Write chunkIdx : %d, file : %s", chunkIdx, file.FileMetadata.Filename)
+	chunk, loaded, err := file.getChunk(chunkIdx)
+	if err == nil && !loaded {
+		close(chunk.ScheduleDownload)
+	}
+	return chunk, err
+}
```

Comment:
since I don't understand ScheduleUpload and ScheduleDownload, I don't understand why we need thise and not just getChunk()